langserver/symbol.go

```diff
@@ -220,10 +212,21 @@ func toSym(name, container string, kind lsp.SymbolKind, fs *token.FileSet, pos t
 	}
 }
 
+// handleTextDocumentSymbol handles `textDocument/documentSymbol` requests for
+// the Go language server.
+func (h *LangHandler) handleTextDocumentSymbol(ctx context.Context, conn JSONRPC2Conn, req *jsonrpc2.Request, params lsp.DocumentSymbolParams) ([]lsp.SymbolInformation, error) {
+	f := strings.TrimPrefix(params.TextDocument.URI, "file://")
+	return h.handleSymbol(ctx, conn, req, query{file: f}, 0)
```

Comment:
I think you meant `collectFromPkg` not `getPackageSyms`.

You're right that `handleSymbol` here iterates over each package invoking `collectFromPkg`, but check out the modification to `collectFromPkg` on lines 295 - 300: it aborts early if the package does not contain the desired file.

It's true that we could _probably_ reduce this further to only operating on a single file (instead of all files in the package the file of interest resides in), but I imagine this would be a lot more involved for not a lot of immediate benefits.